include/genomics/printer.hpp

```diff
@@ -92,67 +95,97 @@ namespace genomics {
+      return cfd;
+    }
+
+    std::tuple<std::string, float> off_target_fields(const kmer& k,
+                                                     genome_structure gs,
+                                                     std::vector<std::vector<std::tuple<int64_t, match>>>& off_targets,
+                                                     int64_t max_off_targets) {
       uint64_t delim = get_delim(gs);
-      std::string out("");
+      std::string offtarget_hex("");
+      float cfd_sum = 0.0;
 
-      for (uint64_t k = 0; k < off_targets.size(); k++) {
+      for (uint64_t i = 0; i < off_targets.size(); i++) {
           std::list<int64_t> v;
 
-          if (max_off_targets != -1) { 
-              std::shuffle(off_targets[k].begin(), off_targets[k].end(), std::mt19937{std::random_device{}()});
-              int64_t count = 0;
-              for (const auto& t : off_targets[k]) {
-                  if (count >= max_off_targets) break;
-                  v.push_back(std::get<0>(t));
-                  count++;
-              }
-          } else {
-            for (const auto& t : off_targets[k]) {
-                v.push_back(std::get<0>(t));
-            }
+          std::shuffle(off_targets[i].begin(), off_targets[i].end(), std::mt19937{std::random_device{}()});
+          int64_t n_off_targets = 0;
+          for (const auto& t : off_targets[i]) {
+              if ((max_off_targets != -1) && (n_off_targets >= max_off_targets))
+                  break;
+              auto pos = std::get<0>(t);
+              auto match = std::get<1>(t);
+
+              coordinates c;
+              std::string strand;
+              tie(c, strand) = resolve_absolute(gs, pos, k);
+              if (c.chr.name == "") continue;
+              v.push_back(pos);
+
+              cfd_sum += calculate_cfd(k.sequence, complement(match.sequence), k.pam);
+
+              n_off_targets++;
           }
 
-          v.push_back(k);
+          v.push_back(i);
           v.push_back(delim);
-          
-          out += list_to_little_endian_hex(v);
+
+          offtarget_hex += list_to_little_endian_hex(v);
       }
 
-      return out;
+      float specificity = 0.0;
+      if (cfd_sum > 0) specificity = 1/cfd_sum;
+
+      return std::make_tuple(offtarget_hex, specificity);
     }
   };
 
   void write_sam_header(std::ostream& os, genome_structure gs) {
     os << "@HD\tVN:1.0\tSO:unknown" << std::endl;
+    os << "@PG\tID:Guidescan\tVN:" << std::string(GUIDESCAN_VERSION) << std::endl;
     for (const auto& chr : gs) {
       os << "@SQ\tSN:" << chr.name << "\tLN:" << chr.length << std::endl;
     }
   }
 
-  void write_csv_header(std::ostream& os) {
-    os << "id,sequence,match_chrm,match_position,match_strand,match_sequence,match_distance,rna_bulges,dna_bulges" << std::endl;
+  void write_csv_header(std::ostream& os, bool complete) {
+    os << "id,sequence,match_chrm,match_position,match_strand,match_distance";
```

Comment:
The csv column order has been tweaked (with corresponding changes in the data for rows) so that all `--complete` columns are bunched. Hope this is okay.